pkg/deployment/config.go

```diff
@@ -116,10 +121,53 @@ func defaultConfig() *Config {
 	return cfg
 }
 
+func (d Config) PodRuntimeClassName(lbs map[string]string) *string {
+	runtimeClassName := ""
+	specificity := -1
+	for k, v := range d.RuntimeClassNames {
+		if !v.Matches(lbs) || v.specificity() < specificity {
+			continue
+		}
+		if v.specificity() > specificity || strings.Compare(k, runtimeClassName) < 0 {
```

Comment:
Is string comparison a rule to select a class when they have equal specificity?
Looks a bit ad hoc. Could you add a test for this and some scenario with multiple runtime classes?

Reply:
This implementation comes from the existing domain selection code for a Service
https://github.com/knative/serving/blob/4fe029fa357276a5a39e912092372a8308aa67e3/pkg/reconciler/route/config/domain.go#L95-L146.

TLDR; the one with most matching labels wins.

Good thinking. I've added two tests for this behaviour in https://github.com/knative/serving/commit/4adbced1bcd13c08e91fd37573dcbddcebc7206e#diff-2a14f9c49eb40c3a55958c8e2d57523cd618232821a9d310c1ddd7cc64d91ef9R476-R517

Reply:
> TLDR; the one with most matching labels wins.

Also a classname can't have a label (key, value)  that does not exist in the service labels.

On top of that, if you have two class names that have the same specificity, lexicographic order in class names is what matters:
("three" < "two"): 
```
{
		name: "priority with multiple label selectors and three labels set",
		serviceLabels: map[string]string{
			"needs-two":   "yes",
			"needs-three": "yes",
			"needs-four":  "yes",
		},
		runtimeClassNames: map[string]RuntimeClassNameLabelSelector{
			"one": {},
			"two": {
				Selector: map[string]string{
					"needs-two":   "yes",
					"needs-three": "yes",
				},
			},
			"three": {
				Selector: map[string]string{
					"needs-two":   "yes",
					"needs-three": "yes",
				},
			},
		},
		want: ptr.String("three"),
	},
```

Reply:
> Also a classname can't have a label (key, value) that does not exist in the service labels.

I'm not sure I understand. Are you meaning to validate a selector against the labels of Services that exist?
If so, that wouldn't make a lot of sense to me.

> On top of that, if you have two class names that have the same specificity, lexicographic order in class names is what matters:
> ("three" < "two"):

Yes that is true, but I would think an uncommon practical issue.

Do you have any alternative patterns for selected which I may consider?

Reply:
> I'm not sure I understand. 

```
		serviceLabels: map[string]string{
			"needs-two":   "yes",
			"needs-three": "yes",
			"needs-four":  "yes",
		},
		runtimeClassNames: map[string]RuntimeClassNameLabelSelector{
			"one": {},
			"two": {
				Selector: map[string]string{
					"needs-two":   "yes",
					"needs-five": "yes",
				},
			},
```
two is ignored due to the v.MatchLabels condition. Matches has this loop that checks against the service label:

```
	for label, expectedValue := range s.Selector {
		value, ok := labels[label]
		if !ok || expectedValue != value {
			return false
		}
	}
```

Reply:
> Do you have any alternative patterns for selected which I may consider?

I am ok as long as we document it.

Reply:
That appears to be correct behaviour, the service labels in the above sample don't include "need-five".

Reply:
@skonto, wonderful!
I've added a line here to explain the behavior: https://github.com/knative/docs/pull/6022/files#diff-55b5f9ddd150abde7c42e2c06362452f26f4d9ce5d902ca5db8c0e27eef518e2R141